### PR TITLE
Correctly support multiple events mapped to a single action

### LINKED ISSUE #30888: Input.is_action_pressed() inconsistent for multiple key bindings
**Godot version:**
3.1.1

**Issue description:**
If assigning multiple keys to the same action, then pressing both keys at the same time and only releasing one (while still holding the second key bound to the same action) will return false for that action in Input.is_action_pressed().
It appears that Input does not check if another key is still being pressed when a key assigned to the same action is released.

**Steps to reproduce:**
-Create a new action in the Project Settings (e.g. "move_up")
-Assign two different keys to that action ("W" and "Up")
-Create a script that checks for Input.is_action_pressed("move_up")
-Press both assigned keys at the same time ("W" and "Up")
-Release only one while still holding the other
-Input.is_action_pressed("move_up") will now return _false_ 

**Minimal reproduction project:**
[Input.is_action_pressed Bugreport.zip](https://github.com/godotengine/godot/files/3438967/Input.is_action_pressed.Bugreport.zip)
The color of the Icon will change based on Input detected for the action "do_stuff". Please press W or Space to trigger the action.

### LINKED ISSUE #39287: InputMap - Action Strength Inconsistency With Key And Analog Stick Mapped
**Godot version:** 3.2.1.stable.official

**OS/device including version:** Microsoft Windows 10 Pro Version	10.0.18363 Build 18363

**Controller:** PS4 Dualshock4

**Issue description:**
Screenshot (6)

When having controller connected, an input event action mapped to a keyboard key (up in project) and a controller analog stick (left stick up in project) will not give a constant strength when the keyboard key is held down.

**Steps to reproduce:**
Open project.
Connect controller.
Press "Up Arrow" Key

**Minimal reproduction project:**

[ControllerStickBug.zip](https://github.com/godotengine/godot/files/4730456/ControllerStickBug.zip)

### LINKED ISSUE #45628: Movement doesn't work while using a controller (d-pad buttons)
**Godot version:**
3.2.4.rc1.official

**OS/device including version:**
Windows 10 (19042.746)

**Issue description:**
When I plug in a PS4 controller, KinematicBody2D doesn't move properly while using the d-pad buttons or the keyboard. Same occurs if I use DS4Windows to emulate a Xbox 360 controller (XInput).

**(!) Note** The same code works fine in Godot v3.2.3.stable, both for the PS4 controller (with and without DS4Windows) and the keyboard.

```
func _physics_process(delta):
	velocity = get_direction() * speed
	velocity = move_and_slide(velocity)


func get_direction():
	var dir: = Vector2.ZERO
	dir.x = Input.get_action_strength("move_right") - Input.get_action_strength("move_left")
	dir.y = Input.get_action_strength("move_down") - Input.get_action_strength("move_up")
	dir = dir.normalized()
	return dir
```

**Steps to reproduce:**
1. Open and run the attached project with a PS4 controller connected.
2. Move the rectangle on the screen using WASD or the arrow keys, then try to move it using the d-pad buttons.
3. The node's movement shouldn't work properly, neither with the keyboard or the controller.
4. Disconnect the controller and the rectangle will move without problems using the keyboard.

Repeat steps 1 and 2 in Godot v3.2.3.stable.official and the rectangle will move without problems.

> [Here](https://streamable.com/45xsjm) you can see how the gamepad registers the buttons pressing without moving the rectangle, same for WASD and arrow keys.

**Minimal reproduction project:**

[GamepadInputError.zip](https://github.com/godotengine/godot/files/5903578/GamepadInputError.zip)

### PR DESCRIPTION
**Pull request by madmiraal**
Currently, when multiple events are mapped to a single action, the state of the action is based on the last event that changed. This PR stores the state of each event to enable determining the correct state of the action regardless of which event last changed. This ensures that when two or more events are mapped to a single action, the Input `is_action_pressed()` reports true if any of the events are still pressed, and `get_action_strength()` returns the maximum value provided by all the events.

Properly fixes #30888 (reverts #30890).
Fixes #39287
Fixes #45628

~Note: I've tested this with keys and mouse buttons, but I've marked this as `needs testing`, because I don't have a joystick, so I can't test the `JoyButton` or `JoyAxis` events, specifically the combination of WASD keys and a joystick described in #39287 and #45628.~

**Edit:** I have now personally tested this with a controller to confirm that it does fix #39287 and #45628.
**Note:** It now includes automatically updating projects' input settings (i.e. it updates a project's `project.godot` file including the `config_version`). This means that projects opened with this PR cannot be opened with an earlier version of Godot (without restoring the `project.godot` file).

### CODE REVIEW

**Comment by madmiraal** [comment]
Rebased following merge of #35666.

**Comment by madmiraal** [comment]
Updated to fix issues identified by @mphe, and rebased to incorporate the changes made to `Input` since this PR was created. It now includes automatically updating projects' input settings (i.e. it updates a project's `project.godot` file including the `config_version`).

I have now personally tested this with a controller to confirm that it does fix #39287 and #45628.

**Comment by jitspoe** [comment]
At a quick glance, the concept looks good.  Does this handle Input.action_press() + Input.action_release() for code-triggered actions?

**Comment by madmiraal** [comment]
> Does this handle Input.action_press() + Input.action_release() for code-triggered actions?

This PR doesn't affect the behaviour of `Input.action_press()` or `Input.action_release()`.

File: core/input/input_event.h [hunk 1]
```diff
@@ -71,18 +71,20 @@ class InputEvent : public Resource {
 	float get_action_strength(const StringName &p_action, bool p_exact_match = false) const;
 	float get_action_raw_strength(const StringName &p_action, bool p_exact_match = false) const;
 
-	// To be removed someday, since they do not make sense for all events
-	virtual bool is_pressed() const;
+	// To be removed someday, since it doesn't make sense for all events
```
Comment (groud) [inline core/input/input_event.h:74]: This comment was for is_pressed I suppose ? Not sure it still make sense here.

**Review by groud — COMMENTED** [review]
I have forgotten a little bit how this area of the codebase works, but I am not fully convinced by the implementation. I seems that `copy_action_values` is kind of a workaround to store the values in the event we store in the input map, but IMO, this is very confusing.

I feel that, if we need to store the action status, we need to augment the input map structure to store things alongside the event, not store things directly in the event itself.
Ultimately, I wonder if we should store the event at all, and use a dedicated structure instead, but that's likely another issue.

On a side note, I think that moving JoyAxisRange to an enum makes sense.

File: core/input/input_event.h (hunk 1 above)
Comment (madmiraal) [reply core/input/input_event.h:74]: The comment was for `is_pressed()` and `is_echo()`, not it's just for `is_echo()`.

**Comment by madmiraal** [comment]
> I seems that copy_action_values is kind of a workaround to store the values in the event we store in the input map, but IMO, this is very confusing.
> I feel that, if we need to store the action status, we need to augment the input map structure to store things alongside the event, not store things directly in the event itself.

What do you suggest as a better way to determine the status of each `Action`? I believe we need to keep a record of the last value associated with each `InputEvent` associated with each `Action`; so it's status can be determined when needed. Since, we're already storing the `InputEvent`s associated with each `Action` in the `input_map`. This PR simply uses this to also store each `InputEvent`'s last value.

**Comment by groud** [comment]
> Since, we're already storing the InputEvents associated with each Action in the input_map. This PR simply uses this to also store each InputEvent's last value.

And yeah, that's exactly the issue I have with the implementation. I don't think we should modify the InputEvent API only for our needs in the action mapping system.

The mapping from event to action does not necessarily need all data contained in the InputEvents we store. For example, we don't need the "pressed" status of an InputEventKey event to map a key press to an action, we would only need a key ID and probably a few other fields. But everything events provide aren't really needed for mapping an event to an action.

To be more precise, I think that, ultimately, we should modify this struct (in input_map.h) by removing `Ref<InputEvent>` and replacing it by a custom structure:
```
	struct Action {
		int id;
		float deadzone;
		List<Ref<InputEvent>> inputs;
	};
```

Anyway, to simplify things for now, I think we should keep `Ref<InputEvent>` there for now, but, instead of storing values directly inside the event. So we should have something like that:
```
        struct EventMapping {
                // The values you need to store
                bool pressed;
                //etc, ...
		Ref<InputEvent> event;
        }
	struct Action {
		int id;
		float deadzone;
		List<EventMapping> inputs;
	};
```
That way, we don't bloat the InputEvent API and keep the fix where it is needed.

**Comment by madmiraal** [comment]
The only thing I've added to the `InputEvent` API is the method `copy_action_values()` and separating the duplicate use of `axis_value` also being used for `axis_range`. Beyond that this PR actually simplifies the `InputEvent` API, by removing the secondary effect of `action_match()` sometimes also retrieving those same action values. This simplification will also allow a follow up PR to further simplify the API by merging `action_match()` and `is_match()`. I wouldn't call this _"modify the InputEvent API only for our needs in the action mapping system."_ never mind _"bloat the InputEvent API"_.

This PR is about properly supporting multiple `InputEvent`s mapped to a single `Action`. It's done using the data structures that are already there and hence _"keep the fix where it is needed."_. Whether or not the data structures should _"eventually"_ be changed is a separate question.

**Comment by dvclmn** [comment]
I am having the same issue — more than happy to provide test information etc, if needed to help resolve. I am on both Windows and MacOS, w/ Dualsense controller and keyboard set to same inputs (with goal to support seamless/automatic switching between different input devices for players). Let me know if I can be of help.

**Comment by Calinou** [comment]
@dvclmn Please keep the discussion about the issue itself in https://github.com/godotengine/godot/issues/45628. This pull request is unlikely to be merged due to the implementation not being up to par.

**Comment by mubinulhaque** [comment]
> @dvclmn Please keep the discussion about the issue itself in #45628.
> 
> This pull request is unlikely to be merged due to the implementation not being up to par. It needs to be redone in a simpler way at this point, while also being rebased on the latest `master` branch.

What does a simpler way mean exactly? I have some time this summer to try doing a PR so I need to know specifics.

**Comment by akien-mga** [comment]
Superseded by #80859.

**Pull request closed without merging**